### PR TITLE
[RFC][ROCm][AITER] Keep all AITER kernels in `_aiter_ops` class like `_custom_ops` and `_ipex_ops`

### PR DESCRIPTION
**Pull request by vllmellm**
## Purpose

This PR introduces `_aiter_ops.py` as proposed in the [RFC here](https://github.com/vllm-project/vllm/issues/21504). The `aiter_ops` namespace provides several key benefits:

- Centralized kernel registration: Ensures that kernels from the aiter package are properly registered

- Environment availability checks: Encapsulates aiter support detection and environment compatibility validation

- Reduced code duplication: Eliminates the need for duplicate helper functions, namely checking device compability and environment varible enability checks across different vLLM modules.

This implementation establishes the foundation for future refactoring efforts, where existing kernels throughout the vLLM repository will be migrated to use this unified approach for better maintainability and consistency.

This PR uses [5ee37dce](https://github.com/ROCm/aiter/commit/5ee37dced6f1bde0229b2c77ce079433549aa25f) commit from `aiter` repo.

## Test Plan

Test models that are afftected by this change, using lm_eval on gsm8k dataset.

**environment setting**

Step 1: run vllm serve

`
VLLM_USE_V1=1 \
VLLM_ROCM_USE_AITER=1 \
SAFETENSORS_FAST_GPU=1 \
VLLM_DISABLE_COMPILE_CACHE=1 \
vllm serve $MODEL_NAME --compilation-config '{"cudagraph_mode": "FULL_AND_PIECEWISE", "cudagraph_capture_sizes": [1,2,4,8,16,24,32]}'  --trust-remote-code --swap-space 16 --distributed-executor-backend mp`

Step 2: run lm_eval

`lm_eval --model local-completions
--tasks gsm8k
--model_args model=$MODEL_NAME,base_url=http://localhost:8000/v1/completions
--trust_remote_code
--num_fewshot 5
--batch_size 256`

### deepseek-ai/DeepSeek-V3 -tp 8 --block-size 1  --max-model-len 32768 --max_seq_len_to_capture 32768 

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9500|±  | 0.006|
|     |       |strict-match    |     5|exact_match|↑  |0.9507|±  | 0.006|

### meta-llama/Llama-4-Scout-17B-16E-Instruct -tp 8 --max-model-len 8192 

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9174|±  |0.0076|
|     |       |strict-match    |     5|exact_match|↑  |0.8999|±  |0.0083|

### meta-llama/Llama-4-Maverick-17B-128E-Instruct-FP8 -tp 8 --max-model-len 8192 

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9272|±  |0.0072|
|     |       |strict-match    |     5|exact_match|↑  |0.9287|±  |0.0071|

### serve Qwen/Qwen3-235B-A22B-FP8 -tp 4

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.8704|±  |0.0093|
|     |       |strict-match    |     5|exact_match|↑  |0.8635|±  |0.0095|

### mistralai/Mixtral-8x7B-Instruct-v0.1 -tp 2

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.6452|±  |0.0132|
|     |       |strict-match    |     5|exact_match|↑  |0.6422|±  |0.0132|

### mistralai/Mixtral-8x7B-Instruct-v0.1 -tp 2 --quantization fp8

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.6058|±  |0.0135|
|     |       |strict-match    |     5|exact_match|↑  |0.6027|±  |0.0135|

### meta-llama/Meta-Llama-3.3-70B-Instruct -tp 2

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9386|±  |0.0066|
|     |       |strict-match    |     5|exact_match|↑  |0.9128|±  |0.0078|

### meta-llama/Meta-Llama-3.3-70B-Instruct -tp 2 --quantization fp8

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9424|±  |0.0064|
|     |       |strict-match    |     5|exact_match|↑  |0.9128|±  |0.0078|

### amd/Llama-3.3-70B-Instruct-FP8-KV -tp 2

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9378|±  |0.0067|
|     |       |strict-match    |     5|exact_match|↑  |0.9030|±  |0.0082|

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

**Comment by vllmellm** [comment]
CC: @ProExpertProg, @robertgshaw2-redhat @DarkLight1337

**Comment by ProExpertProg** [comment]
Could you resolve conflicts?

**Comment by tjtanaa** [comment]
Hi, @maleksan85 , could you help to review the FP4 related logic?

File: vllm/model_executor/layers/quantization/utils/fp8_utils.py [hunk 1]
```diff
@@ -68,171 +67,6 @@ def cutlass_scaled_mm(
...
-
-
-direct_register_custom_op(
-    "w8a8_triton_block_scaled_mm_func",
-    _w8a8_triton_block_scaled_mm_func,
-    fake_impl=_w8a8_triton_block_scaled_mm_fake,
-)
-
-
-def _padded_cutlass(
```
Comment (ProExpertProg) [inline vllm/model_executor/layers/quantization/utils/fp8_utils.py:179]: Where did padded_cutlass go?

File: vllm/model_executor/layers/quantization/utils/fp8_utils.py [hunk 2]
```diff
@@ -68,171 +67,6 @@ def cutlass_scaled_mm(
...
-        import aiter as rocm_aiter
-        from aiter import get_hip_quant
-
-        aiter_per1x128_quant = get_hip_quant(rocm_aiter.QuantType.per_1x128)
-
-
-# TODO we should be able to change the type of block_size to GroupShape
-# after we resolve GroupShape compilation issue
-# https://github.com/vllm-project/vllm/issues/25270
-def _w8a8_triton_block_scaled_mm_func(
```
Comment (ProExpertProg) [inline vllm/model_executor/layers/quantization/utils/fp8_utils.py:146]: Where did this custom op go?

File: vllm/model_executor/layers/quantization/quark/schemes/quark_ocp_mx.py [hunk 3]
```diff
@@ -31,6 +31,13 @@
 logger = init_logger(__name__)
 
 
+# TODO: move registration of custom op to aiter_ops.py
+# `from vllm._aiter_ops import rocm_aiter_ops`
+# use `rocm_aiter_ops.is_asm_fp4_gemm_dynamic_quant_enabled()`
+# for envs checks which does not require @cache anymore.
+# triton kernel is torch compile compatible.
+# does not require direct registeration.
+# use `rocm_aiter_ops.triton_fp4_gemm_dynamic_qaunt`.
```
Comment (ProExpertProg) [inline vllm/model_executor/layers/quantization/quark/schemes/quark_ocp_mx.py:40]: I don't understand: what is the blocker for this TODO? Why can't we do it in this PR?

File: vllm/model_executor/layers/fused_moe/layer.py [hunk 4]
```diff
@@ -1575,20 +1578,19 @@ def _map_global_expert_id_to_local_expert_id(self, expert_id: int) -> int:
...
-                    tp_rank=self.ep_rank if self.use_ep else self.tp_rank,
-                    tp_size=self.ep_size if self.use_ep else self.tp_size,
-                    shared_experts_score=1.0,
-                    max_num_tokens=vllm_config.scheduler_config.max_num_batched_tokens
-                    * dp_size,
-                    is_EP=self.use_ep,
-                )
-            self.local_num_experts += self.num_fused_shared_experts
+        if self.num_fused_shared_experts > 0:
+            init_aiter_topK_meta_data(
```
Comment (ProExpertProg) [inline vllm/model_executor/layers/fused_moe/layer.py:1760]: I know this is not new to this PR but I'm curious: where does this meta_data get stored? In a global variable?

File: vllm/_aiter_ops.py [hunk 5]
```diff
@@ -0,0 +1,899 @@
...
+from vllm.platforms import current_platform
+from vllm.utils.torch_utils import direct_register_custom_op, is_torch_equal_or_newer
+
+
+def is_aiter_supported(func: Callable) -> Callable:
+    """Decorator that only executes the function if
+    ROCm AITER package is supported on gfx9 archs.
+    """
+
+    @wraps(func)
```
Comment (ProExpertProg) [inline vllm/_aiter_ops.py]: Nit: fully qualified name
```suggestion
    @funcutils.wraps(func)
```

File: vllm/_aiter_ops.py [hunk 6]
```diff
@@ -0,0 +1,899 @@
...
+from functools import wraps
+
+import torch
+
+import vllm.envs as envs
+from vllm.platforms import current_platform
+from vllm.utils.torch_utils import direct_register_custom_op, is_torch_equal_or_newer
+
+
+def is_aiter_supported(func: Callable) -> Callable:
```
Comment (ProExpertProg) [inline vllm/_aiter_ops.py]: Call this `if_aiter_supported`

**Review by ProExpertProg — COMMENTED** [review]
Apologies for the delay, this is great cleanup, thanks for this contribution! I think there were a couple of non-aiter deletions, please double-check the code.

File: vllm/model_executor/layers/quantization/utils/fp8_utils.py (hunk 2 above)
Comment (vllmellm) [reply vllm/model_executor/layers/quantization/utils/fp8_utils.py:146]: @ProExpertProg thanks for pointing out i think by mistake it was removed while resolving merge conflict. They are recovered back in.

File: vllm/model_executor/layers/quantization/quark/schemes/quark_ocp_mx.py (hunk 3 above)
Comment (vllmellm) [reply vllm/model_executor/layers/quantization/quark/schemes/quark_ocp_mx.py:40]: @ProExpertProg The TODO is suggesting to move the aiter kernel used in this module to `aiter_ops.py` and the instruction how to do it like how to check enablement of the aiter kernel is used here as well the aiter triton kernel exists in `aiter_ops.py` that can be replaced without doing additional direct_resigter on vllm library since in this module I see that the triton kernel is combined with aiter asm kernel and both together are registered while triton kernel is torch.compile compatible and doesnt require to be registered using  `direct_register_custom_ops` . we cant unfortunately apply the changes since we cant test module is for a device architecture that our team dont have access for testing. so we would need original author to apply the changes after this RFC.

File: vllm/model_executor/layers/fused_moe/layer.py (hunk 4 above)
Comment (vllmellm) [reply vllm/model_executor/layers/fused_moe/layer.py:1760]: @ProExpertProg Yes it is stored in a global variable and the reason behind it is to avoid memory leaks introducing the the buffer in this metadata causes memory usage overhead if done for each layer so this global variable avoids such case. Do you have any other suggestion for this?

**Comment by vllmellm** [comment]
@ProExpertProg The missing codes are recovered and addressed your previous comments. would appreciate your help in reviewing the PR again.

**Comment by tjtanaa** [comment]
@SageMoore could you take a look as well thank you.

File: vllm/model_executor/layers/layernorm.py [hunk 7]
```diff
@@ -169,11 +118,14 @@ def __init__(
             self.weight = nn.Parameter(self.weight)
 
         if current_platform.is_rocm():
+            aiter_rmsnorm_enabled = rocm_aiter_ops.is_rmsnorm_enabled()
```
Comment (SageMoore) [inline vllm/model_executor/layers/layernorm.py:121]: Nit. Let's move this line into the `dispatch_rocm_rmsnorm_func`

**Review by SageMoore — COMMENTED** [review]
What a PR. Great work. This is an enormous step in the right direction and is a significant cleanup of the AITER integration. I think a really good next step would be to do AITER op enablement similar to custom op enablement on the command line and get rid of these environment variables, but that can be a follow on. This PR lays a good foundation for that.

File: vllm/model_executor/layers/layernorm.py (hunk 7 above)
Comment (vllmellm) [reply vllm/model_executor/layers/layernorm.py:121]: @SageMoore Thanks for the comment and review.
this function is wrapped by a decorator that checks device spec and whether `aiter` package is installed using `find_spec(aiter)` which this function if not torch.compile compatible. since dispatch function is called during forward pass and when torch.compile is enabled we get error with torch  dynamo not being able to trace the path.

Comment (ProExpertProg) [reply vllm/model_executor/layers/layernorm.py:121]: Let's go the other way then, and not dispatch during the forward pass and do all dispatching ahead of time

Comment (vllmellm) [reply vllm/model_executor/layers/layernorm.py:121]: @ProExpertProg  yes most of the aiter kernels availability checks are done elsewhere ahead of time like in  `__init__` functions except in fused_moe layer when selecting experts by `fused_topk` functions in `select_experts` (a method in `FusedMoE` class) since this method is a `staticmethod` then we have the enability checked within the body of this `select_experts`. So to avoid any breaks, I added a new commit that checks `aiter` package installation globally in `_aiter_ops.py` as below.

```
def is_aiter_found() -> bool:
    from importlib.util import find_spec

    return find_spec("aiter") is not None


# `find_spec` is not torch.compile compatible.
# In cases where aiter availability might have
# been checked in forward passes that are torch compiled.
# we keep this global outside to not cause torch compile breaks.
IS_AITER_FOUND = is_aiter_found()


def if_aiter_supported(func: Callable) -> Callable:
    """Decorator that only executes the function if
    ROCm AITER package is supported on gfx9 archs.
    """

    @functools.wraps(func)
    def wrapper(*args, **kwargs):
        # checks the platform, device arch and aiter library existance.

        from vllm.platforms.rocm import on_gfx9

        if current_platform.is_rocm() and on_gfx9() and IS_AITER_FOUND:
            return func(*args, **kwargs)
        else:
            # Return None or do nothing if not supported
            return None

    return wrapper
```

Comment (ProExpertProg) [reply vllm/model_executor/layers/layernorm.py:121]: Wait I did not see the context, the line this comment refers to is inside `RMSNorm.__init__` which is not a part of the forward pass. As @SageMoore said we should be able to move `dispatch_rocm_rmsnorm_func` into `dispatch_rocm_rmsnorm_func` instead of calling outside and passing it as a param?

Comment (vllmellm) [reply vllm/model_executor/layers/layernorm.py:121]: @ProExpertProg sorry I got confused as well. So we have unit tests for all the dispatching logic to make sure it is returning the intended function. when the envs variable check is inside the dispatch functions we need some hacky ways to clear cache and reload aiter_ops module since the envs variable are cached when aiter_ops is imported due to loading the envs as class variables in aiter_ops. During unit test we know that if writing a unit test becomes complicated to just test a simple function means the design is wrong. by making the dispatcher function independent from env variable checks while making them to return intended function solely based on passed arguments makes it easy to write test for them. 
on the other hand, we are working on PR that implements aiter_rmsnorm fusion pass and dispatcher is no longer needed. I guess for this PR for now we can keep this code as it is? if it's ok to both of you @ProExpertProg @SageMoore. Thanks again

**Review by SageMoore — APPROVED** [review]

**Review by tjtanaa — APPROVED** [review]
